### PR TITLE
[WIP] rpi-kernel: build flavor subpackages, add rpi4

### PR DESCRIPTION
**Pull request by Piraty**
This now builds all Raspberry Pi kernels from a single template by using
individual builddir/destdir for each flavor and performing all stages
independently with help of the `_flavor_stage()` function.
    
 * rpi-kernel is an empty meta package now that is built by default for
   every arch and is kept to serve legacy installations as it depends on
   the appropriate new flavor rpi{1,2,3}-kernel package.
 * subpackages are enabled conditionally for each arch.
 * add rpi4 while at it :-)

see: https://github.com/void-linux/void-mklive/pull/153 as well

---

### How to
Build packages
```
cd void-packages
./xbps-src -m masterdir.rpi4 binary-bootstrap
./xbps-src -m masterdir.rpi4 -a aarch64-musl pkg rpi-base
./xbps-src -m masterdir.rpi4 -a aarch64-musl pkg rpi-kernel
```
Build images: see https://github.com/void-linux/void-mklive/pull/153

---

### Remaining issues
- [ ] change to `rpi-base` require manual action for existing instals (`rpi{0,1,2,3}`). (rpi-kernel will result in an orphan) is there a better way?

### CODE REVIEW

**Comment by liketechnik** [comment]
(I hope it's appropriate to ask here)
I'd love to help testing, but I'm not sure how/what.

I'm currently running void (64bit) on a rpi4 (currently put void's kernel packages on hold an manually copied the latest release from the rpi firmware repository). I also have a rpi3 available, although I'd need some more time to buy an SD-card for it and set the system up on it.

What would be the next steps for me? (e. g. what pre-requisites do I need [local repository to compile the packages?] and what would I need to change on my rpi [which packages would be updated/installed new)

File: srcpkgs/rpi-kernel/template
Comment (agausmann) [inline srcpkgs/rpi-kernel/template:71]: These dependencies are also set for the flavor packages, causing for example `rpi4-kernel` to depend on `rpi3-kernel`, since `rpi4-kernel` is built for `aarch64`. This is definitely unwanted, and causes problems with `./mkplatformfs.sh rpi4 ...` when it attempts to install both `rpi3-kernel` and `rpi4-kernel`, as the files will conflict.

File: srcpkgs/rpi-kernel/template
Comment (agausmann) [inline srcpkgs/rpi-kernel/template:342]: Same issue as on [line 71](#diff-7c06a56d03c82f4f91449e0b8baba00fa54433735eee8b84183aa75909a25b34R71)

**Comment by agausmann** [comment]
Tested on my rpi1, builds and boots to a login shell successfully. USB, video and networking all seem to be fine; didn't expect much to change there, but just for the sake of completeness.

Also tried to test on rpi4, but ran into issues while building, which I have detailed above.

File: srcpkgs/rpi-kernel/template
Comment (Piraty) [reply srcpkgs/rpi-kernel/template:71]: oh right. will fix it. transition handling is not ideal yet

Comment (Piraty) [reply srcpkgs/rpi-kernel/template:71]: i cannot confirm. `depends=` is not inherited by subpakages. the toplevel declaration you refereced only affects the meta packages, see below (notice `run_depends`)
```
$ XBPS_TARGET_ARCH=aarch64-musl xbps-query --repository=$HOME/src/void-packages/hostdir/binpkgs -M -R rpi4-kernel                                                                                       
architecture: aarch64-musl
filename-sha256: 422a8449148b60cf6dd53f80f61f2b556b4bb505cc6588e6c5877721e6273054
filename-size: 28MB
homepage: http://www.kernel.org
installed_size: 39MB
license: GPL-2.0-only
maintainer: Piraty <piraty1@inbox.ru>
pkgname: rpi4-kernel
pkgver: rpi4-kernel-5.4.68_1
repository: /home/piraty/src/void-packages/hostdir/binpkgs
short_desc: Linux kernel for Raspberry Pi 4 (5.4 [git e82816d])

$ XBPS_TARGET_ARCH=aarch64-musl xbps-query --repository=$HOME/src/void-packages/hostdir/binpkgs -M -R rpi-kernel
architecture: aarch64-musl
filename-sha256: 0686ecf163d116935cac4c2f744ff3d00e2baf023ff952cd4c4c143ba4752a24
filename-size: 529B
homepage: http://www.kernel.org
installed_size: 0B
license: GPL-2.0-only
maintainer: Piraty <piraty1@inbox.ru>
pkgname: rpi-kernel
pkgver: rpi-kernel-5.4.68_1
repository: /home/piraty/src/void-packages/hostdir/binpkgs
run_depends:
        rpi3-kernel>=0
short_desc: Linux kernel for Raspberry Pi (5.4 series [git e82816d])
```

File: srcpkgs/rpi-kernel/template
Comment (Piraty) [reply srcpkgs/rpi-kernel/template:342]: that doesn't make sense. this is set on the subpackage only, i don't see how rpi3-kernel-headers would pull rpi4-kernel-headers

File: srcpkgs/rpi-kernel/template
Comment (agausmann) [reply srcpkgs/rpi-kernel/template:71]: Yes, I get the same result that you do. However this is still somehow a problem, even if this isn't the source. When I run `./xbps-src -a aarch64 pkg rpi4-kernel`, for some reason. it also builds `rpi-kernel`, `rpi-kernel-headers`, and transitively, `rpi3-kernel` and `rpi3-kernel-headers`. And then, for some reason, `./mkplatformfs.sh rpi4 ...` attempts to install them.

```
$ ls hostdir/binpkgs/rpi-kernel-flavor-subpackage/
aarch64-repodata                          rpi3-kernel-headers-5.4.68_1.aarch64.xbps
rpi-kernel-5.4.68_1.aarch64.xbps          rpi4-kernel-5.4.68_1.aarch64.xbps
rpi-kernel-headers-5.4.68_1.aarch64.xbps  rpi4-kernel-headers-5.4.68_1.aarch64.xbps
rpi3-kernel-5.4.68_1.aarch64.xbps
```
see also: the [build log](https://termbin.com/9he0z) for `./xbps-src -a aarch64 pkg rpi4-kernel`, which builds both rpi3-kernel and rpi4-kernel.

**Comment by manneorama** [comment]
Currently trying to test this rebased on top of the current void-packages master (on top of `f66d93e241420fbb17f8c1fffadf897bba676382` to be specific). Kernel package builds fine but during building of base-system I run into a strange error while building `elfutils`. 

```
Making all in debuginfod
  CXXLD    debuginfod
/usr/bin/ld: ../libelf/libelf.so: undefined reference to symbol 'inflateEnd'
/usr/bin/ld: /lib64/libz.so.1: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:581: debuginfod] Error 1
make[1]: *** [Makefile:502: all-recursive] Error 1
make: *** [Makefile:418: all] Error 2
=> ERROR: elfutils-0.182_4: do_build: '${make_cmd} ${makejobs} ${make_build_args} ${make_build_target}' exited with 2
=> ERROR:   in do_build() at common/build-style/gnu-configure.sh:13
```

I'm not exactly sure where to start digging, but I will do my best. Any help is appreciated.

**Comment by manneorama** [comment]
I finally gave in and simply built everything on an existing rpi4 installation instead of going the cross compile route. With the small modification I mentioned in the `void-mklive` PR I could successfully build an image and boot off of it. 

I don’t know if my host setup or the cross compilation in `xbps-src` is broken for rpi4, but something is wrong at least. I was able to get around the original issue by just starting over (?). But then I had other issues and at one point the compiler somehow disappeared which is when I gave up. I will have another, more structured, go at the cross compilation once I’m done testing the image I now have.

**Comment by Anachron** [comment]
Has someone replaced their RPI 3 B+ install with this PR yet? I'm thinking about copying the sdcard and modifying the install here and there to not need to reinstall and reconfigure everything.

**Comment by LeamHall** [comment]
Noob question; is there a mailing list or IRC channel I can get on to watch this? my RPi 4 should be here tomorrow.  :)

**Comment by manneorama** [comment]
I've been running this on my Pi 4B's for a few days now and so far it's rock solid. All my pi's are headless so I have not tested graphical output, but networking (wired and wireless) and everything else seems to work as expected.

**Comment by CMB** [comment]
I built this and then made an image for my Pi 4 using the void-mklive
pull request, updating the case statement for selecting the kernel,
as discussed in that PR.  It appears to be working quite well.

File: srcpkgs/rpi-base/template
Comment (ahesford) [inline srcpkgs/rpi-base/template:4]: Revbump for the dependency drop

**Review by ahesford — APPROVED** [review]
This looks good to me. I took my existing 3b+ setup based on an rpi3 PLATFORMFS image, updated `rpi-base`, dropped `rpi-kernel` and installed `rpi4-kernel`. I've been able to boot this image from a USB stick as expected.

File: srcpkgs/rpi-base/template
Comment (ericonr) [inline srcpkgs/rpi-base/template:6]: I still think current users shouldn't have to change anything about their current systems to get things working. The amount of information contained in this package is small enough that a `rpi4-base` package wouldn't be undue maintenance in almost any way.

Comment (jsumners) [reply srcpkgs/rpi-base/template:6]: What current systems? Void does not currently support RPI4 without a lot of manual work.

Comment (ericonr) [reply srcpkgs/rpi-base/template:6]: Current users of the `rpi-base` packages, so rpi zero, 1, 2 and 3 owners.

Comment (jsumners) [reply srcpkgs/rpi-base/template:6]: Those are technically different architectures. I wouldn't think it wise to try and run an install targeting any of those on any other device.

Comment (ericonr) [reply srcpkgs/rpi-base/template:6]: I'm not sure I get what you mean.

My point is that `rpi-base` currently works and pulls all dependencies for current rpi users. With this change, they will also have to download the kernel for their specific board. It would be simpler for them, without much more work on our side, to have a `rpi4-base` package for rpi4 users (it could even be a subpackage here), while current users don't get an orphaned kernel package or have to remember to install new systems with `rpi-base` *and* `rpi$something-kernel`.

Comment (ahesford) [reply srcpkgs/rpi-base/template:6]: I don't like the idea of a separate `rpi4-base` because the distinction is arbitrary. The problem compounds when there is a Raspberry Pi 5 and so on. Suddenly every new hardware release that requires a different kernel gets paired with a new `-base` just to pull in that kernel.

The only difference needed to move from rpi3 to rpi4 is the kernel, so let's decouple the kernel from the base and keep things consistent.

Comment (Piraty) [reply srcpkgs/rpi-base/template:6]: that was my point, yet i didn't come up with a solution that doesn't require user intervention for existing install (1rpi{0,1,2,3}1)

Comment (ahesford) [reply srcpkgs/rpi-base/template:6]: Having per-model bases seems better than playing "one of these things is not like the other" with `rpi-base` and `rpi4-base`. I don't think there's a lot of added value in making the `-base` depend on the kernel, but if there is agreement with the position of @ericonr, might as well split them all if we're splitting one.

Comment (ericonr) [reply srcpkgs/rpi-base/template:6]: For what it's worth, rpi0 and rpi1 would be the same package.

File: srcpkgs/rpi-kernel/template
Comment (ahesford) [inline srcpkgs/rpi-kernel/template:220]: This block causes problems because some dtb files are missing in `rpi3-kernel` and `rpi4-kernel` (the same problem affects the current `rpi3-kernel` package). For example, on my 3b+ which tries to use NFS root with kernel IP configuration, the Ethernet adapter gets a new, random MAC address with every boot. The `smsc95xx.macaddr` kernel command-line argument makes no difference. The problem is resolved by copying `/boot/broadcom/bcm2710-rpi-3-b-plus.dtb` and `/boot/broadcom/bcm2837-rpi-3-b-plus.dtb` to `/boot`.

I think the right thing here is to copy all of the relevant family and SoC dtb files (see [this table](https://openwrt.org/toh/raspberry_pi_foundation/raspberry_pi#hardware)):

```bash
case "$rpi_flavor in
	rpi3)
		mv ${DESTDIR}/boot/broadcom/bcm{2710,2837}-*.dtb ${DESTDIR}/boot
		;;
	rpi4)
		mv ${DESTDIR}/boot/broadcom/bcm2711-*.dtb ${DESTDIR}/boot
		;;
esac
rm -rf ${DESTDIR}/broadcom
```
This should capture all of the proper DTBs for `rpi3`-class hardware. For `rpi4`, there is currently only one DTB, but it looks like there are more `bcm2711-*` DTBs for the RPi 400 and CM4 [coming in 5.10.x](https://github.com/raspberrypi/linux/tree/rpi-5.10.y/arch/arm/boot/dts), so doing a glob probably saves us some trouble later.

File: srcpkgs/rpi-kernel/template
Comment (CameronNemo) [reply srcpkgs/rpi-kernel/template:71]: @agausmann the behavior you described for building subpackages is expected. You build a source package, and all of its subpackages get built. If there are issues with how mkplatformfs handles things, then we can iron it out in the PR over there.

File: srcpkgs/rpi-kernel/template
Comment (Piraty) [reply srcpkgs/rpi-kernel/template]: good catch @ahesford , thank you

**Comment by Piraty** [comment]
to sum up current state of discussion:
- is building all flavors in the rpi-kernel template fine with everyone? (split into disinct templates could be done later anyway).
  - pro: one commit per update
  - con: unnecessarily long build times if i just need rpi3 but build rpi4 as well in the same run
- as with the current patch, rpi-kernel will become an orphan for installed rpi{0,1,2,3} systmes but pull in the new rpi{1,2,3}-kernel. an appropriate INSTALL.msg is provided
- how to handle rpi0 (as pointed out by @ericonr [here](https://github.com/void-linux/void-packages/pull/26000#discussion_r555040991))? just implicitly assume to run rpi1-kernel?
- better split rpi-base into rpi{1,2,3,4}-base as well? that might resolve the rpi-kernel orphan situation as the transitional dummy rpi-base could pull rpi{1,2,3}-base as per target arch mapping which ensures the dependency chain is kept intact
  -

@the-maldridge @pbui @ahesford @ericonr @void-linux/pkg-committers

**Comment by ahesford** [comment]
I think it's OK for the Zero to require `rpi1-kernel` because it's a first-generation RPi SoC. By the same token, I believe a few later RPi 2 designs should be able to run`rpi3-kernel` (with the dtb copy fix, anyway) because those are effectively third-generation SoCs in the second-generation form factor (not that we should advertise this---sufficiently knowledgeable users can find their way).

I don't have a strong opinion about splitting `rpi-base`. As @ericonr points out, this would let people simply pick the right base to grab what they need. (Should these depend on one of `base-{minimal,voidstrap,system}` to provide a more proper "base"?) On the other hand, the only difference is in the kernel dependency, so the HOWTO isn't made much more complicated by telling users to install `rpi-base` plus the kernel for their hardware generation.

**Comment by ericonr** [comment]
> is building all flavors in the rpi-kernel template fine with everyone? (split into disinct templates could be done later anyway).

maybe make a build option (or one build option for each) to enable either one, if possible/simple? That way the day-to-day maintenance remains simple, but someone wanting to build locally can select just one.

> how to handle rpi0 (as pointed out by @ericonr here)? just implicitly assume to run rpi1-kernel?

Yes. Just make the `short_desc` for `rpi1-base` and `rpi1-kernel` mention "raspberrypi zero and 1" or similar).

> better split rpi-base into rpi{1,2,3,4}-base as well? that might resolve the rpi-kernel orphan situation as the transitional dummy rpi-base could pull rpi{1,2,3}-base as per target arch mapping which ensures the dependency chain is kept intact

This is my preference.

**Comment by Bridouz** [comment]
So until this PR is merged we should kept 'linux-rpi' back on Rpi 4 ('rootfs' images found on reddit) right ?

**Comment by jsumners** [comment]
> So until this PR is merged we should kept 'linux-rpi' back on Rpi 4 ('rootfs' images found on reddit) right ?

I do not know what images you refer to, but I most certainly _would not_ trust any image floating around on Reddit. If you are having issues building this branch and need a pre-built image, you would be better off trusting my image found at https://github.com/void-linux/void-packages/pull/23541#issuecomment-683442628 . It's not _much_ better than trusting some random image on Reddit, but at least I took care to publish it via as trustworthy a medium as I could.

**Comment by paper42** [comment]
5.10 is now available for Raspberry Pi. Should it be updated in a separate PR or here?

**Comment by jsav0** [comment]
Confirmed cross-compiling and then booting two RPI4s 4GB models here, no issues but the kernel is 5.4.68_1. When can we expect this to be officially merged?

**Comment by Piraty** [comment]
after i took some long thought on this topic, i'll close this PR and pursue https://github.com/void-linux/void-packages/pull/29139 instead.
Reasoning: keeping templates split for each kernel appears to be a cleaner solution and might be beneficial for (future future) kernel build style, as well as build times when building one of rpi3 or rpi4

**Pull request closed without merging**
